### PR TITLE
JDK-8315706: com/sun/tools/attach/warnings/DynamicLoadWarningTest.java real fix for failure on AIX

### PR DESCRIPTION
**Pull request by JoKern65**
After push of [JDK-8307478](https://bugs.openjdk.org/browse/JDK-8307478) , the following test started to fail on AIX :
com/sun/tools/attach/warnings/DynamicLoadWarningTest.java;
The problem was described in [JDK-8309549](https://bugs.openjdk.org/browse/JDK-8309549) with a first try of a fix.
A second fix via [JDK-8310191](https://bugs.openjdk.org/browse/JDK-8310191) was necessary.
Both fixes just disable the specific subtest on AIX, without correction of the root cause.
The root cause is, that dlopen() on AIX returns different handles every time, even if you load a library twice. There is no official AIX API available to get this information on a different way.
My proposal is, to use the stat64x API with the fields st_device and st_inode. After a dlopen() the stat64x() API is called additionally to get this information which is then stored parallel to the library handle in the jvmtiAgent. For AIX we then can compare these values instead of the library handle and get the same functionality as on linux.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8315706](https://bugs.openjdk.org/browse/JDK-8315706): com/sun/tools/attach/warnings/DynamicLoadWarningTest.java real fix for failure on AIX (**Bug** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Matthias Baesken](https://openjdk.org/census#mbaesken) (@MBaesken - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15583/head:pull/15583` \
`$ git checkout pull/15583`

Update a local copy of the PR: \
`$ git checkout pull/15583` \
`$ git pull https://git.openjdk.org/jdk.git pull/15583/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15583`

View PR using the GUI difftool: \
`$ git pr show -t 15583`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15583.diff">https://git.openjdk.org/jdk/pull/15583.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15583#issuecomment-1707895997)

### CODE REVIEW

**Comment by AlanBateman** [comment]
Just to point out that AIX behavior is not wrong, the spec is clear that "it is implementation specific as to whether a warning is printed when attempting to start the same agent a second or subsequent time". Just pointing it out as you have the option to avoid this complexity if you want. The cited JBS issues are because the original tests for JEP 451 assumed that there would be additional warnings when the same agent was loaded/started several times.

**Comment by JoKern65** [comment]
Why not having a solution for AIX anyway?

File: test/jdk/com/sun/tools/attach/warnings/DynamicLoadWarningTest.java [hunk 1]
```diff
@@ -123,8 +123,6 @@ void testLoadOneJvmtiAgent(OnAttachAction loadJvmtiAgent1) throws Exception {
                 .whenRunning(loadJvmtiAgent1)
                 .stderrShouldNotContain(JVMTI_AGENT_WARNING);
 
-        // test behavior on platforms that can detect if an agent library was previously loaded
-        if (!Platform.isAix()) {
```
Comment (dholmes-ora) [inline test/jdk/com/sun/tools/attach/warnings/DynamicLoadWarningTest.java:127]: You need to fix the indentation of the old block.

**Comment by dholmes-ora** [comment]
I would much rather see an AIX solution that was in the AIX version of `os::dll_load` rather than having to pollute the shared JVMTI code. I'm not sure how best to achieve that - it may not be possible to hide it completely - but we should be able to refactor things so `stat64x_LIBPATH` is defined in AIX code, and its use is via a helper so the code is only written once. Then we would only need a handful of `AIX_ONLY(...)` statements.

File: src/hotspot/share/prims/jvmtiAgentList.cpp [hunk 2]
```diff
@@ -243,6 +243,19 @@ bool JvmtiAgentList::is_dynamic_lib_loaded(void* os_lib) {
   }
   return false;
 }
+#ifdef AIX
+bool JvmtiAgentList::is_dynamic_lib_loaded(long device, long inode) {
+  JvmtiAgentList::Iterator it = JvmtiAgentList::agents();
+  while (it.has_next()) {
+    JvmtiAgent* const agent = it.next();
+    if (!agent->is_static_lib() && device && inode &&
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/jvmtiAgentList.cpp:251]: Style nit: we don't use implicit booleans so check `device != 0` and  `inode != 0` explicitly please.

Comment (JoKern65) [reply src/hotspot/share/prims/jvmtiAgentList.cpp]: I followed your suggestion

File: test/jdk/com/sun/tools/attach/warnings/DynamicLoadWarningTest.java (hunk 1 above)
Comment (JoKern65) [reply test/jdk/com/sun/tools/attach/warnings/DynamicLoadWarningTest.java:127]: I followed your suggestion here.

**Comment by JoKern65** [comment]
I moved `stat64x_LIBPATH` to AIX code, and tried to use some `AIX_ONLY(...)` statements. I hope this is better.

**Review by dholmes-ora — COMMENTED** [review]
That is a definite improvement - thanks.

Lets see what others think.

File: src/hotspot/share/prims/jvmtiAgent.hpp [hunk 3]
```diff
@@ -43,6 +43,10 @@ class JvmtiAgent : public CHeapObj<mtServiceability> {
   const char* _options;
   void* _os_lib;
   const char* _os_lib_path;
+#ifdef AIX
+  long _inode;
+  long _device;
```
Comment (AlanBateman) [inline src/hotspot/share/prims/jvmtiAgent.hpp:48]: How are dev_t and ino_t defined on AIX, I'm wondering if long is okay here.

**Review by AlanBateman — COMMENTED** [review]
No objection to doing this but just to repeat again that the spec does not not require that warnings are de-duplicated. This means the change is not necessary, only happens the case where someone is loading the same agent dynamically many times into a VM running on AIX, it just means there will be one rather than N warnings in the logs.

File: src/hotspot/share/prims/jvmtiAgent.hpp (hunk 3 above)
Comment (JoKern65) [reply src/hotspot/share/prims/jvmtiAgent.hpp:48]: They are defined as __ulong64_t which is unsigned long. So I can change it to unsigned long or even to dev_t and ino_t.

File: src/hotspot/os/aix/os_aix.cpp [hunk 4]
```diff
@@ -3043,3 +3043,32 @@ void os::print_memory_mappings(char* addr, size_t bytes, outputStream* st) {}
...
+  if (env == nullptr || *env == 0)
+    return -1;
+
+  int ret = -1;
+  size_t libpathlen = strlen(env);
+  char* libpath = NEW_C_HEAP_ARRAY(char, libpathlen + 1, mtServiceability);
+  char* combined = NEW_C_HEAP_ARRAY(char, libpathlen + strlen(path) +1, mtServiceability);
+  char *saveptr, *token;
+  strcpy(libpath, env);
+  for( token = strtok_r(libpath, ":", &saveptr); token != nullptr; token = strtok_r(nullptr, ":", &saveptr) ) {
```
Comment (RealCLanger) [inline src/hotspot/os/aix/os_aix.cpp:3065]: spaces:
```suggestion
  for (token = strtok_r(libpath, ":", &saveptr); token != nullptr; token = strtok_r(nullptr, ":", &saveptr)) {
```

File: src/hotspot/os/aix/os_aix.cpp [hunk 5]
```diff
@@ -3043,3 +3043,32 @@ void os::print_memory_mappings(char* addr, size_t bytes, outputStream* st) {}
...
+  }
+
+  const char* env = getenv("LIBPATH");
+  if (env == nullptr || *env == 0)
+    return -1;
+
+  int ret = -1;
+  size_t libpathlen = strlen(env);
+  char* libpath = NEW_C_HEAP_ARRAY(char, libpathlen + 1, mtServiceability);
+  char* combined = NEW_C_HEAP_ARRAY(char, libpathlen + strlen(path) +1, mtServiceability);
```
Comment (RealCLanger) [inline src/hotspot/os/aix/os_aix.cpp:3062]: Space between +1
```suggestion
  char* combined = NEW_C_HEAP_ARRAY(char, libpathlen + strlen(path) + 1, mtServiceability);
```

File: src/hotspot/share/prims/jvmtiAgent.cpp [hunk 6]
```diff
@@ -272,6 +294,21 @@ static bool load_agent_from_executable(JvmtiAgent* agent, const char* on_load_sy
...
+#ifdef AIX
+// save the inode and device of the library's file as a signature. This signature can be used
+// in the same way as the library handle as a signature on other platforms.
+static void save_library_signature(JvmtiAgent* agent, const char* name) {
+  struct stat64x libstat;
+  if (0 == os::Aix::stat64x_via_LIBPATH(name, &libstat)) {
+    agent->set_inode(libstat.st_ino);
+    agent->set_device(libstat.st_dev);
+  }
+  else {
```
Comment (RealCLanger) [inline src/hotspot/share/prims/jvmtiAgent.cpp:306]: Fix style: 
```suggestion
  } else {
```

**Review by RealCLanger — COMMENTED** [review]
Just a few indentation remarks.

Could you please also update the SAP copyright year in src/hotspot/os/aix/os_aix.hpp?

**Comment by JoKern65** [comment]
Again, I followed the proposals.

File: src/hotspot/share/prims/jvmtiAgent.hpp [hunk 7]
```diff
@@ -43,6 +43,10 @@ class JvmtiAgent : public CHeapObj<mtServiceability> {
   const char* _options;
   void* _os_lib;
   const char* _os_lib_path;
+#ifdef AIX
+  unsigned long _inode;
+  unsigned long _device;
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/jvmtiAgent.hpp:48]: It is best, IMO, to use the actual types rather than something expected to be "equivalent".

**Review by dholmes-ora — CHANGES_REQUESTED** [review]

File: src/hotspot/share/prims/jvmtiAgent.hpp (hunk 7 above)
Comment (JoKern65) [reply src/hotspot/share/prims/jvmtiAgent.hpp:48]: OK, this would be ino64_t and dev64_t. I will do the change.

**Review by dholmes-ora — APPROVED** [review]
Nothing further from me. Thanks.

**Review by MBaesken — APPROVED** [review]

**Comment by tstuefe** [comment]
This now causes problems with 

https://github.com/openjdk/jdk/pull/16604#issuecomment-1827661214

since it removes the possibility of silently alternating the file path inside os::dll_load, which would be the preferred way for AIX to handle *.a shared objects. So this causes even more ifdef AIX to bloom up everywhere.

A much better solution would have been to mimic stable-handle behavior inside the AIX version of `os::dll_load`. 

Proposal for an alternate solution: Hold dlhandle-to-(inode, devid)tuple mappings in a hash table. On dlopen, look up dl-handle by (inode, filename) tupel. On dlclose, remove entry. Could have been done inside os_aix.cpp without any changes to shared coding, and would have provided complete coverage for all users of dll_load.

**Comment by JoKern65** [comment]
@tstuefe: Hi Thomas, I'm not sure if I got it. We can make (inode, devid) to a hash, which replaces the dlhandle on return of os::dlload. This hash would of course be the same if the same library is loaded twice. But I do not know how to handle the two real dlhandles.

**Comment by tstuefe** [comment]
> @tstuefe: Hi Thomas, I'm not sure if I got it. We can make (inode, devid) to a hash, which replaces the dlhandle on return of os::dlload. This hash would of course be the same if the same library is loaded twice. But I do not know how to handle the two real dlhandles.

Why do you need two dlhandles? 

A handle returned from dlopen should be valid for the whole process. If you cache that in a hashmap, and for the second caller of os::dll_load() return the cached variant, this should work, no?

**Comment by JoKern65** [comment]
If you have time, please call me for a short discussion.

**Pull request closed without merging**